receiver/sqlserverreceiver/scraper.go

```diff
@@ -300,3 +329,228 @@ func (s *sqlServerScraperHelper) recordDatabaseStatusMetrics(ctx context.Context
 
 	return errors.Join(errs...)
 }
+
+func (s *sqlServerScraperHelper) recordDatabaseQueryTextAndPlan(ctx context.Context, topQueryCount uint) (plog.Logs, error) {
+	// Constants are the column names of the database status
+	const totalElapsedTime = "total_elapsed_time"
+	const rowsReturned = "total_rows"
+	const totalWorkerTime = "total_worker_time"
+	const queryHash = "query_hash"
+	const queryPlanHash = "query_plan_hash"
+	const logicalReads = "total_logical_reads"
+	const logicalWrites = "total_logical_writes"
+	const physicalReads = "total_physical_reads"
+	const executionCount = "execution_count"
+	const totalGrant = "total_grant_kb"
+	const queryText = "query_text"
+	const queryPlan = "query_plan"
+	rows, err := s.client.QueryRows(ctx)
+	if err != nil {
+		if errors.Is(err, sqlquery.ErrNullValueWarning) {
+			s.logger.Warn("problems encountered getting metric rows", zap.Error(err))
+		} else {
+			return plog.Logs{}, fmt.Errorf("sqlServerScraperHelper failed getting rows: %w", err)
+		}
+	}
+	var errs []error
+
+	totalElapsedTimeDiffs := make([]int64, len(rows))
+
+	for i, row := range rows {
+		queryHashVal := hex.EncodeToString([]byte(row[queryHash]))
+		queryPlanHashVal := hex.EncodeToString([]byte(row[queryPlanHash]))
+
+		elapsedTime, err := strconv.ParseInt(row[totalElapsedTime], 10, 64)
+		if err != nil {
+			s.logger.Info(fmt.Sprintf("sqlServerScraperHelper failed getting rows: %s", err))
+			errs = append(errs, err)
+		} else {
+			// we're trying to get the queries that used the most time.
+			// caching the total elapsed time (in millisecond) and compare in the next scrape.
+			if cached, diff := s.cacheAndDiff(queryHashVal, queryPlanHashVal, totalElapsedTime, elapsedTime/1000); cached && diff > 0 {
+				totalElapsedTimeDiffs[i] = diff
+			}
+		}
+	}
+
+	// sort the rows based on the totalElapsedTimeDiffs in descending order
+	rows = sortRows(rows, totalElapsedTimeDiffs)
+
+	// sort the totalElapsedTimeDiffs in descending order as well
+	sort.Slice(totalElapsedTimeDiffs, func(i, j int) bool { return totalElapsedTimeDiffs[i] > totalElapsedTimeDiffs[j] })
+
+	logs := plog.NewLogs()
+
+	for i, row := range rows {
+		if i >= int(topQueryCount) {
+			break
+		}
+
+		// we will hit this in two cases:
+		//	(1) not cached as first run
+		//	(2) the corresponding query is not taking more CPU time
+		if totalElapsedTimeDiffs[i] == 0 {
+			continue
+		}
+
+		// reporting human-readable query hash and query hash plan
+		queryHashVal := hex.EncodeToString([]byte(row[queryHash]))
+		queryPlanHashVal := hex.EncodeToString([]byte(row[queryPlanHash]))
+
+		record := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
```

Comment:
If the exported logs don't have the `scope` from https://github.com/open-telemetry/opentelemetry-proto/blob/d9fd878d46b1970cf72b3d11ea012fff60716aa8/opentelemetry/proto/logs/v1/logs.proto#L72C33-L72C53, which contains `scope.name` and `scope.version`, we may need to populate these values manually.

I am not sure whether the collector has a built-in behavior to populate this field for every receiver or not.